typing/typemod.ml

```diff
@@ -707,6 +708,77 @@ let merge_constraint initial_env loc sg lid constr =
   with Includemod.Error explanation ->
     raise(Error(loc, initial_env, With_mismatch(lid.txt, explanation)))
 
+(* A simplified version of [merge_constraint] for the case of packages. Package
+   constraints are much simpler - they must be "with type" constraints for types
+   with no parameters and can only update abstract types. *)
+let merge_package_constraint initial_env loc sg lid cty =
+  let rec patch_item namelist outer_sig_env sg_for_env ~ghosts item =
+    let return replace_by =
+      Some ((), {Signature_group.ghosts; replace_by})
+    in
+    match item, namelist with
+    | Sig_type(id, sig_decl, rs, priv) , [s]
+      when Ident.name id = s ->
+        begin match sig_decl.type_manifest with
+        | None -> ()
+        | Some ty ->
+          raise (Error(loc, outer_sig_env, With_package_manifest (lid.txt, ty)))
+        end;
+        let new_sig_decl =
+          Typedecl.transl_package_constraint ~loc outer_sig_env cty.ctyp_type
+        in
+        check_type_decl outer_sig_env sg_for_env loc id None
+          new_sig_decl sig_decl;
+        let new_sig_decl = { new_sig_decl with type_manifest = None } in
+        return (Some(Sig_type(id, new_sig_decl, rs, priv)))
+    | Sig_module(id, _, md, rs, priv) as item, s :: namelist
+      when Ident.name id = s ->
+        let sig_env = Env.add_signature sg_for_env outer_sig_env in
+        let sg = extract_sig sig_env loc md.md_type in
+        let ((), newsg) = merge_signature sig_env sg namelist in
+        let item =
+          match md.md_type with
+            Mty_alias _ ->
+              (* A module alias cannot be refined, so keep it
+                 and just check that the constraint is correct *)
+              item
+          | _ ->
+              let newmd = {md with md_type = Mty_signature newsg} in
+              Sig_module(id, Mp_present, newmd, rs, priv)
```

Comment:
I am wondering if we want to share this logic for module aliases with `merge_constraint` by having a common `update_module` function.